sdk/webpubsub/azure-messaging-webpubsub-client/src/main/java/com/azure/messaging/webpubsub/client/implementation/websocket/WebSocketClientHandler.java

```diff
@@ -138,4 +198,53 @@ public CompletableFuture<Void> getClientCloseCallbackFuture() {
     CloseWebSocketFrame getServerCloseWebSocketFrame() {
         return this.serverCloseWebSocketFrame;
     }
+
+    private static final class OpenTextFrame {
+        private final ArrayList<ByteBuf> collected = new ArrayList<>();
+        private final AtomicBoolean isClosed = new AtomicBoolean(false);
+
+        private volatile BinaryData collectedData;
+
+        private void add(ByteBuf buffer) {
+            if (isClosed.get()) {
+                throw new IllegalStateException("Cannot add more buffers when closed.");
+            }
+
+            buffer.retain();
+            collected.add(buffer);
+        }
+
+        private void close() {
+            if (isClosed.getAndSet(true)) {
+                getCollectedData();
+            }
+        }
+
+        private void dispose() {
+            if (isClosed.get()) {
+                collected.forEach(b -> b.release());
+                collected.clear();
+            }
+        }
+
+        private BinaryData getCollectedData() {
```

Comment:
Would it be simpler, if this 3 functions is merged together? It would mark itself as closed, collect data, and close all ByteBuf. There would be no concern about whether the user of this class forget any of them.

On the concurrency, do we need to do all these protections?
I haven't done the full research, but it seems netty won't call `channelRead0` concurrently.
https://github.com/netty/netty/wiki/New-and-noteworthy-in-4.0#well-defined-thread-model